enaml/src/subscription_observer.cpp

```diff
@@ -0,0 +1,332 @@
+        return 0;
+    }
+    self->name =  cppy::incref( name );
+    return ptr.release();
+}
+
+
+void
+SubscriptionObserver_clear( SubscriptionObserver* self )
+{
+    Py_CLEAR( self->atomref );
+    Py_CLEAR( self->name );
+}
+
+
+int
+SubscriptionObserver_traverse( SubscriptionObserver* self, visitproc visit, void* arg )
+{
+    Py_VISIT( self->atomref );
+    Py_VISIT( self->name );
+    Py_VISIT(Py_TYPE(self));
+    return 0;
+}
+
+
+void
+SubscriptionObserver_dealloc( SubscriptionObserver* self )
+{
+    PyObject_GC_UnTrack( self );
+    SubscriptionObserver_clear( self );
+    Py_TYPE(self)->tp_free( reinterpret_cast<PyObject*>( self ) );
+}
+
+
+int
+SubscriptionObserver__bool__( SubscriptionObserver* self )
+{
+    return PyObject_IsTrue( self->atomref );
+}
+
+
+/*
+ * Calls engine update with the owner and name
+ * if self.ref:
+ *     owner = self.ref()
+ *     engine = owner._d_engine
+ *      if engine is not None:
+ *         engine.update(owner, self.name)
+ */
+PyObject*
+SubscriptionObserver_call( SubscriptionObserver* self, PyObject* args, PyObject* kwargs )
+{
+
+    if( PyObject_IsTrue( self->atomref ) )
+    {
+        cppy::ptr owner( PyObject_CallNoArgs( self->atomref ) );
+        cppy::ptr engine( owner.getattr("_d_engine") );
+        if ( !engine.is_none() )
+        {
+            cppy::ptr update_args( PyTuple_New( 2 ) );
+            if( !update_args )
+                return 0;
+            PyTuple_SET_ITEM( update_args.get(), 0, cppy::incref( owner.get() ) );
+            PyTuple_SET_ITEM( update_args.get(), 1, cppy::incref( self->name ) );
+            cppy::ptr update( engine.getattr( "update" ) );
+            return update.call( update_args );
+        }
+    }
+    Py_RETURN_NONE;
+}
+
+
+PyObject*
+SubscriptionObserver_richcompare( SubscriptionObserver* self, PyObject* other, int opid )
+{
+    if( opid == Py_EQ )
+    {
+        if( SubscriptionObserver::TypeCheck( other ) )
+        {
+            SubscriptionObserver* so_other = reinterpret_cast<SubscriptionObserver*>( other );
+            if(
+                PyObject_RichCompareBool( self->atomref, so_other->atomref, Py_EQ )
+                && PyObject_RichCompareBool( self->name, so_other->name, Py_EQ )
+            )
+            {
+                Py_RETURN_TRUE;
+            }
+        }
+        Py_RETURN_FALSE;
+    }
+    Py_RETURN_NOTIMPLEMENTED;
+}
+
+
+PyDoc_STRVAR(SubscriptionObserver__doc__,
+    "SubscriptionObserver(owner, name)\n\n"
+    "An observer object which manages a tracer subscription.\n"
+    "Parameters\n"
+    "----------\n"
+    "owner : Declarative\n"
+    "    The declarative owner of interest.\n\n"
+    "name : string\n"
+    "    The name to which the operator is bound\n");
+
+
+PyObject*
+SubscriptionObserver_get_ref( SubscriptionObserver* self, void* context )
+{
+    return cppy::incref( self->atomref );
+}
+
+
+PyObject*
+SubscriptionObserver_set_ref( SubscriptionObserver* self, PyObject* value, void* context )
```

Comment:
I think the only time this is ever called to be set is with `None`. I think we can just check that `value == Py_NONE` here and throw an exception otherwise.

Reply:
Agree, cppy:replace seems to do both in one go.